File: algorithms/CopyListWithRandomPointer/CopyListWithRandomPointer.py
class Solution(object):
    def copyRandomList(self, head):
        d = {}
        node = head
        while node:
            d[node] = RandomListNode(node.label)
            node = node.next

        node = head
        while node:
            d[node].next = d.get(node.next)
            d[node].random = d.get(node.random)
            node = node.next
        return d.get(head)

File: algorithms/CopyListWithRandomPointer/CopyListWithRandomPointer.cpp
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        map<RandomListNode*, RandomListNode*> m;
        for(RandomListNode* node = head; node; node = node->next) {
            m[node] = new RandomListNode(node->label);
        }
        for(RandomListNode* node = head; node; node = node->next) {
            m[node]->next = m[node->next];
            m[node]->random = m[node->random];
        }
        return m[head];
    }
};
